keps/sig-windows/0000-20181221-windows-group-managed-service-accounts-for-container-identity.md

```diff
@@ -0,0 +1,225 @@
+  - Provision the service account and gives details to application admin.
+  - Assign access to a security group to control what machines can use this service account. This group should include all authorized Kubernetes nodes.
+
+2. A Kubernetes admin will create a namespace, configmap and service account:
+  - Create a unique Kubernetes [service account](https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/#use-multiple-service-accounts) as an identity for the application in Kubernetes. No rights to the cluster are assigned.
+
+```
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: webserver
+```
+
+  - Run a Windows PowerShell script to verify the account exists, and capture enough data to uniquely identify the GMSA into a JSON file (credspec). This doesn't contain any passwords or crypto secrets.
+
+Example credspec.json
+```
+{
+  "CmsPlugins": [
+    "ActiveDirectory"
+  ],
+  "DomainJoinConfig": {
+    "DnsName": "contoso.com",
+    "Guid": "244818ae-87ca-4fcd-92ec-e79e5252348a",
+    "DnsTreeName": "contoso.com",
+    "NetBiosName": "DEMO",
+    "Sid": "S-1-5-21-2126729477-2524075714-3094792973",
+    "MachineAccountName": "WebApplication1"
+  },
+  "ActiveDirectoryConfig": {
+    "GroupManagedServiceAccounts": [
+      {
+        "Name": "WebApplication1",
+        "Scope": "DEMO"
+      },
+      {
+        "Name": "WebApplication1",
+        "Scope": "contoso.com"
+      }
+    ]
+  }
+}
+```
+
+  - Create a configmap in the namespace with the contents of the JSON - `kubectl create configmap webserverCredSpec --from-file=gmsa-cred-spec=/path/to/credspec.json`
+  - Add reference to the configmap in the service account - `kubectl patch serviceaccount/webserver -p '{\"windowsCredentialSpec\": \"webserverCredSpec\"}'`
+3. The Kubernetes admin gives a set of users rights to the namespace and service account
+4. Application admin deploys the app with `ServiceAccount: webserver` as part of the PodSpec
+5. The ServiceAccount admission controller validates this user has access to the service account, and sets `PodSpec.SecurityContext.WindowsGMSACredentialSpec` to the contents of the credspec configmap
+6. The Windows CRI implementation copies the credentialSpec to the [OCI windows.CredentialSpec](https://github.com/opencontainers/runtime-spec/blob/master/config-windows.md#credential-spec) field
+7. The Windows OCI implementation validates the credentialspec, and fails to start the container if its invalid or access to the GMSA from the node is denied
+8. The container starts with the Windows Group Managed Service Account as expected, and can authenticate to the database.
+
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### Changes in ServiceAccount struct
+A new field `WindowsCredentialSpec String` will be added to `ServiceAccount` to point to the name of a config map with key `gmsa-cred-spec` and data populated with the credspec JSON corresponding to the GMSA.
```

Comment:
If someone wanted to do the same thing for Kerberos, would we then add another field here?

What’s special about the binding of a service account to a Windows domain user that is distinct from any other identity provider?  Why this api structure vs alternatives?

Some alternative examples as discussion points:

* a WindowsServiceAccountBinding object that the Windows kubelet code reads
* an annotation on the service account that contains the credential spec and admission control that prevents it being set
* an opaque identity object reference that can have a group, kind, and name that points to an arbitrary object on or off kube, that the Windows kubelet code can read,”?

Reply:
The GMSA setup allows for no secret material to be kept on the cluster itself.  AD is responsible for holding the credentials and then determining what identities have access to it.  If we wanted to do something similar for Kerberos we would need to consider using a K8s secret.

Reply:
Can this be prototyped with an annotation? I like Clayton's option 3, but I don't want to design that here. I don't want to see fields for every possible integration (GMSA, AWS Role, GCP Service Account, etc...) added to the ServiceAccount object.

Reply:
Why isn't Kerberos like gMSA?  You don't have to have a keytab for a node level thing to grant the workload an identity the node is authorized to act as.